lib/server.js

```diff
@@ -0,0 +1,131 @@
+}
+
+/** Makes an instance of docker exec running command, then links the provided
+    websocket to the stream provided by docker exec
+
+    @param tty: boolean representing whether or not terminal emulation is active
+  */
+function execWithSocket (container, socket, command, tty) {
+  var options = {
+    AttachStdout: true,
+    AttachStderr: true,
+    AttachStdin: true,
+    Tty: tty,
+    Detach: false,
+    Cmd: command,
+  };
+  var attachoptions = {
+    stdin: true,
+    stdout: true,
+    stderr: true,
+    stream: true,
+  };
+  return container.exec(options, (err, exec) => {
+    if (err) { //err proof
+      throw err;
+    }
+    exec.start(attachoptions, (err, execStream) => {
+      if (err) { //err proof
+        throw err;
+      }
+      var strout = through((data) => {
+        if (dbgstream) {
+          dbgstream.write('OUT: ' + data + '\n-\n');
+        }
+        socket.send(muxBuffer(1, data), {binary: true, mask: true});
+      });
+      var strerr = through((data) => {
+        if (dbgstream) {
+          dbgstream.write('ERR: ' + data + '\n-\n');
+        }
+        socket.send(muxBuffer(2, data), {binary: true, mask: true});
+      });
+      exec.modem.demuxStream(execStream, strout, strerr);
+
+      socket.on('message', (message) => {
+        if (dbgstream) {
+          dbgstream.write('IN: ' + message + '\n-\n');
+        }
+        execStream.write(message);
+      });
+
+      socket.on('disconnect', () => {
+        execStream.end('\x03\x04\r\nexit\r\n'); //should be how to ctrl+c ctrl+d, might be better way to kill
+      });
+
+      execStream.on('end', () => {
+        exec.inspect((err, data) => {
+          if (err) {
+            throw err;
+          }
+          var exitBuf = new Buffer(4);
+          exitBuf.writeUInt32LE(data.ExitCode);
+          socket.send(muxBuffer(100, exitBuf));
+        });
+      });
+
+      socket.send('ready'); //send ready message to let client know it is ready
+    });
+  });
+}
+
+export default class DockerExecWebsocketServer {
+  /* Creates Docker Exec instance on given container, running the first message given
+   * as a command.
+   * Options:
+   * port, required
+   * container, name or id of docker container, required
+   * path, path where the websocket is hosted
+   * tty, whether or not we expect VT100-style output
+   */
+  constructor (options) {
+    options = _.defaults({}, options, {path: '/'+slugid.v4(), tty: false});
+    var container = docker.getContainer(options.container);
+    this.port = options.port;
+    this.path = options.path;
+    var wsopts = {
+      port: this.port,
+      path: this.path,
+    };
+    assert(this.port, 'required port option missing');
+    assert(container, 'required container option missing');
+    this.server = new ws.Server(wsopts);
+    debug(wsopts.port+wsopts.path+' created');
+    debug(this.server.options.path);
+    this.server.on('connection', (socket) => {
```

Comment:
I dived a bit into `ws` and it seems that we can get the request that was upgraded to a websocket from `websocket.upgradeReq`. Hence, we can put options like `tty` in querystring... Maybe even the `command` to run in querystring, wouldn't that make the flow easier?

```js
var url = require('url');
var args = url.parse(websocket.upgradeReq.url, true).query;
var tty = (args.tty === 'true');
var cmd = args.cmd;
```

I suspect this will reduce the number of special messages we need to send. What do you think...